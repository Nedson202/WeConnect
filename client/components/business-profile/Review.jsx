import 'rc-pagination/assets/index.css';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import ImageZoom from 'react-medium-image-zoom';
import ReactStars from 'react-stars';
import avatar from '../../images/user-avatar.png';
import EditReview from './EditReview';

/**
 * @class BusinessList
 * @extends {Component}
 */
class Review extends Component {
  /**
   * @description Display buttons
   * @returns {undefined}
   * @memberof ReviewList
   */
  displayButtons() {
    const {
      username,
      review,
      edit,
      deleteStatus,
    } = this.props;
    const { reviewer } = review;
    if (username === reviewer.username || username === process.env.ADMIN) {
      return (
        <div>
          <button
            className="close text-danger review-delete"
            data-toggle="modal"
            data-target="#confirmModal"
            onClick={() => deleteStatus(review.id)}
          >
            <i className="fa fa-trash" />
          </button>
          <button
            onClick={edit(review)}
            className="close"
          >
            <i className="fa fa-edit" />
          </button>
        </div>
      );
    }
    if (username !== reviewer.username && username !== process.env.ADMIN) {
      return null;
    }
    if (username === reviewer.username && username !== process.env.ADMIN) {
      return null;
    }
    if (username === reviewer.username || username === process.env.ADMIN) {
      return (
        <div>
          <button
            className="close text-danger"
            data-toggle="modal"
            data-target="#confirmModal"
          >
            <i className="fa fa-trash" />
          </button>
        </div>
      );
    }
  }

  /**
   * @description Renders the component to the dom
   *
   * @returns {object} JSX object
   *
   * @memberof ReviewList
   */
  render() {
    const {
      review,
      params,
      fetchReviews,
      reviewUpdateRequest,
      fetchBusinessById,
      cancelEdit,
      state,
      onReviewDelete,
      onChange,
      onStarClick,
      onReviewUpdate
    } = this.props;
    const {
      reviewer, message, rating, id, updatedAt
    } = review;

    const { image } = reviewer;

    if (state.editStatus === id) {
      return (
        <div>
          { !reviewer.image ? <img src={avatar} className="rounded-circle user-avatar" alt="Avatar" width="50" height="50" />
          : <img src={reviewer.image} className="rounded-circle user-avatar" alt="Avatar" width="50" height="50" />
        }
          <div className="review-body">
            <span className="textcase reviewer">{reviewer.username} </span>
            <EditReview
              reviewId={id}
              params={params}
              reviewUpdateRequest={reviewUpdateRequest}
              fetchReviews={fetchReviews}
              cancelEdit={cancelEdit}
              fetchBusinessById={fetchBusinessById}
              onChange={onChange}
              state={state}
              onStarClick={onStarClick}
              onReviewUpdate={onReviewUpdate}
            />
          </div>
        </div>
      );
    }

    return (
      <div>
        { !image ? <img src={avatar} className="rounded-circle user-avatar" alt="Avatar" width="50" height="50" />
          :
        <ImageZoom
          image={{
              src: `${image}`,
              alt: 'business image',
              className: 'img',
              style: {
 width: '50px', height: '50px', marginRight: '15px', float: 'left', borderRadius: '50%'
}
            }}
          zoomImage={{
              src: `${image}`,
              alt: 'business image',
            }}
        />
        }
        <div className="review-body">
          <span className="textcase reviewer">{reviewer.username} </span>
          {
            this.displayButtons()
          }
          <div id={id}>
            <p>{message}</p>
            <ReactStars
              count={5}
              edit={false}
              half
              size={17}
              color1="#333333"
              color2="#ffaf00"
              value={rating}
            />
          </div>
          <p
            className="mb-0 small font-weight-medium
            mb-1 text-muted lts-2px"
          >
            {moment(updatedAt).format('MMMM Do YYYY, h:mm:ss a')}
          </p>
        </div>

        <div className="modal fade" id="confirmModal" tabIndex="-1" role="dialog" aria-labelledby="confirmModalTitle" aria-hidden="true">
          <div className="modal-dialog modal-dialog-centered" role="document">
            <div className="modal-content">
              <div className="modal-body">
                <h4>Are you sure?</h4>
              </div>
              <div className="modal-footer">
                <button type="button" className="btn btn-outline-success" data-dismiss="modal">No</button>
                <button type="button" className="btn btn-outline-success" onClick={onReviewDelete(params.id, id)} data-dismiss="modal">Yes</button>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Review.propTypes = {
  fetchBusinessById: PropTypes.func.isRequired,
  reviewUpdateRequest: PropTypes.func.isRequired,
  onReviewDelete: PropTypes.func.isRequired,
  onStarClick: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onReviewUpdate: PropTypes.func.isRequired,
  deleteStatus: PropTypes.func.isRequired,
  edit: PropTypes.func.isRequired,
  cancelEdit: PropTypes.func.isRequired,
  fetchReviews: PropTypes.func.isRequired,
  review: PropTypes.object.isRequired,
  params: PropTypes.object.isRequired,
  state: PropTypes.object.isRequired,
  username: PropTypes.string
};

Review.defaultProps = {
  username: ''
};

export default Review;
