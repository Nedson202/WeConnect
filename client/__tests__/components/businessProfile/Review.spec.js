import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJSON from 'enzyme-to-json';
import Review from '../../../components/business-profile/Review';
import businessData from '../../__mockData__/business';

const { review } = businessData;
configure({ adapter: new Adapter() });

let props;
const setup = () => {
  props = {
    onChange: jest.fn(() => Promise.resolve()),
    onStarClick: jest.fn(() => Promise.resolve()),
    fetchReviews: jest.fn(() => Promise.resolve()),
    onReviewDelete: jest.fn(() => Promise.resolve()),
    onReviewUpdate: jest.fn(() => Promise.resolve()),
    fetchBusinessById: jest.fn(() => Promise.resolve()),
    reviewUpdateRequest: jest.fn(() => Promise.resolve()),
    edit: jest.fn(() => Promise.resolve()),
    cancelEdit: jest.fn(() => Promise.resolve()),
    deleteStatus: jest.fn(() => Promise.resolve()),
    state: {
      review,
      editStatus: true
    },
    review: {
      id: null,
      reviewer: {
        username: 'allen',
        image: '/path/to/image'
      }
    },
    params: {},
    username: 'mike',
    reviewId: 1,
  };

  return shallow(<Review {...props} />);
};


describe('Component: Review', () => {
  beforeEach(() => {
    global.window = {
      window: () => {},
    };
    global.reviewer = {
      image: () => ''
    };
  })

  it('should render', () => {
    const wrapper = setup();
    expect(wrapper.find('div').length).toBe(8);
    expect(wrapper.find('ImageZoom').length).toBe(1);
    expect(wrapper.find('button').length).toBe(2);
    // expect(wrapper.find('EditReview').length).toBe(1);
    wrapper.setProps({
      review: {
        reviewer: {
          username: 'allen',
        }
      }, 
      username: 'allen',
    });
    expect(wrapper.find('img').length).toBe(1);
    expect(wrapper.find('button').length).toBe(4);
    expect(wrapper.find('div').length).toBe(9);
    expect(wrapper.find('span').length).toBe(1);
    wrapper.setProps({
      review: {
        id: 1,
        reviewer: {
          username: 'allen',
        }
      }, 
      username: 'allen',
      state: {
        review,
        editStatus: 1
      },
    });
    expect(wrapper.find('img').length).toBe(1);
    expect(wrapper.find('div').length).toBe(2);
    expect(wrapper.find('span').length).toBe(1);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});

describe('Function test', () => {
  it('should call imageChange function', () => {
    const wrapper = setup();
    const action = wrapper.instance();
    const displayButtons = jest.spyOn(wrapper.instance(), 'displayButtons');
    action.displayButtons();
    expect(displayButtons).toBeCalled();
  });
});