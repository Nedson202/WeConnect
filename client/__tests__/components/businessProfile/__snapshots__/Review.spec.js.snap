// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Review should render 1`] = `
<div>
  <img
    alt="Avatar"
    className="rounded-circle user-avatar"
    height="50"
    src="test-file-stub"
    width="50"
  />
  <div
    className="review-body"
  >
    <span
      className="textcase reviewer"
    >
      allen
       
    </span>
    <EditReview
      cancelEdit={[MockFunction]}
      fetchBusinessById={[MockFunction]}
      fetchReviews={[MockFunction]}
      onChange={[MockFunction]}
      onReviewUpdate={[MockFunction]}
      onStarClick={[MockFunction]}
      params={Object {}}
      reviewId={1}
      reviewUpdateRequest={[MockFunction]}
      state={
        Object {
          "editStatus": 1,
          "review": Object {
            "message": "Great business",
            "rating": 3,
          },
        }
      }
    />
  </div>
</div>
`;
